### ISSUE TITLE
MultiThreaded Script Losing API Connection

### ISSUE THREAD

**Issue by scotthoopes** [cubewise-code/tm1py#890]
**Describe what did you try to do with TM1py**
The basics of the script is I am pulling leaf level data from TM1 into a dataframe then pushing that data into SQL. I am using the following code: 

`df = tm1.cells.execute_mdx_dataframe(mdx=mdxstr)`

Pretty simple call. The dataframes can be anywhere from 200K to 600K rows of data. I am using the threading module to make it so this runs more quickly. It is breaking down the mdx so when it is put into the above code it is looping through a time dimension and another dimension to break down the amount of data in one dataframe. 

We historically have had an on prem version of TM1 and I have been running this code for about 4 months without any issues. When I start this code we have about 100 threads running at a time. Although TM1 tends to accept about 15  connections at a time.

**Describe what's not working the way you expect**
I was expecting that all these threads would run using the provided MDX string and bringing back that data in a dataframe.

What is currently happening: We are currently on PA cloud. Since the change when we run this code about 30% of the threads fail. In command line I am getting the following errors: 

**ChunkedEncodingError: ('Connection broken: IncompleteRead(0 bytes read, 2 more expected)', IncompleteRead(0 bytes read, 2 more expected))**

Full Error below. 

This didn't happen before not sure what would cause it now other than some constraint on the cloud. 

I have checked all the timeouts on the TM1 Config file and they are the same on our old on prem version vs. the cloud.
The whole process normally runs in about 25 minutes. With the errors it runs in about 17 minutes. 
If I run a job that has less threads (about 6 is what I am testing) and that doesn't have any issues. 

Does anyone know if there are thread constraints on PA Cloud? Any other ideas on what might be going wrong?

**Version**
- TM1py [e.g. 1.3.1] 
- TM1 Server Version: PA Cloud Up to Date

**Additional context**
```
F:\Python\lib\site-packages\TM1py\Services\[CellService.py](https://cellservice.py/)", line 50, in wrapper

    return func(self, cellset_id, *args, **kwargs)

  File "F:\Python\lib\site-packages\TM1py\Services\[CellService.py](https://cellservice.py/)", line 2271, in extract_cellset_raw

    cellset_response = self.extract_cellset_raw_response(

  File "F:\Python\lib\site-packages\TM1py\Services\[CellService.py](https://cellservice.py/)", line 2233, in extract_cellset_raw_response

    response = self._rest.GET(url=url, **kwargs)

  File "F:\Python\lib\site-packages\TM1py\Services\[RestService.py](https://restservice.py/)", line 80, in wrapper

    response = self.retrieve_async_response(async_id)

  File "F:\Python\lib\site-packages\TM1py\Services\[RestService.py](https://restservice.py/)", line 590, in retrieve_async_response

    return self._s.get(url, **kwargs)

  File "F:\Python\lib\site-packages\requests\[sessions.py](https://sessions.py/)", line 600, in get

    return self.request("GET", url, **kwargs)

  File "F:\Python\lib\site-packages\requests\[sessions.py](https://sessions.py/)", line 587, in request

    resp = self.send(prep, **send_kwargs)

  File "F:\Python\lib\site-packages\requests\[sessions.py](https://sessions.py/)", line 745, in send

    r.content

  File "F:\Python\lib\site-packages\requests\[models.py](https://models.py/)", line 899, in content

    self._content = b"".join(self.iter_content(CONTENT_CHUNK_SIZE)) or b""

  File "F:\Python\lib\site-packages\requests\[models.py](https://models.py/)", line 818, in generate

    raise ChunkedEncodingError(e)

requests.exceptions.ChunkedEncodingError: ('Connection broken: IncompleteRead(0 bytes read, 2 more expected)', IncompleteRead(0 bytes read, 2 more expected))

Exception in thread Thread-15 (inner_loop1):

Traceback (most recent call last):

  File "F:\Python\lib\site-packages\urllib3\[response.py](https://response.py/)", line 444, in _error_catcher

    yield

  File "F:\Python\lib\site-packages\urllib3\[response.py](https://response.py/)", line 831, in read_chunked

    chunk = self._handle_chunk(amt)

  File "F:\Python\lib\site-packages\urllib3\[response.py](https://response.py/)", line 785, in _handle_chunk

    self._fp._safe_read(2)  # Toss the CRLF at the end of the chunk.

  File "F:\Python\lib\http\[client.py](https://client.py/)", line 632, in _safe_read

    raise IncompleteRead(data, amt-len(data))

http.client.IncompleteRead: IncompleteRead(0 bytes read, 2 more expected)



During handling of the above exception, another exception occurred:



Traceback (most recent call last):

  File "F:\Python\lib\site-packages\requests\[models.py](https://models.py/)", line 816, in generate

    yield from self.raw.stream(chunk_size, decode_content=True)

  File "F:\Python\lib\site-packages\urllib3\[response.py](https://response.py/)", line 624, in stream

    for line in self.read_chunked(amt, decode_content=decode_content):

  File "F:\Python\lib\site-packages\urllib3\[response.py](https://response.py/)", line 816, in read_chunked

    with self._error_catcher():

  File "F:\Python\lib\[contextlib.py](https://contextlib.py/)", line 153, in __exit__

    self.gen.throw(typ, value, traceback)

  File "F:\Python\lib\site-packages\urllib3\[response.py](https://response.py/)", line 461, in _error_catcher

    raise ProtocolError("Connection broken: %r" % e, e)

urllib3.exceptions.ProtocolError: ('Connection broken: IncompleteRead(0 bytes read, 2 more expected)', IncompleteRead(0 bytes read, 2 more expected))



During handling of the above exception, another exception occurred:



Traceback (most recent call last):

  File "F:\Python\lib\[threading.py](https://threading.py/)", line 1009, in _bootstrap_inner

    self.run()

  File "F:\Python\lib\[threading.py](https://threading.py/)", line 946, in run

    self._target(*self._args, **self._kwargs)

  File "F:\Python\Production\tm1_functions.py", line 344, in inner_loop1

    dataframe_base_changes(column_param=week, where_param=s, mdxstr=mdxstr, view_columns=view_columns,

  File "F:\Python\Production\tm1_functions.py", line 325, in dataframe_base_changes

    df = extract_data_base_view(mdxstr, where_param, column_param, scenario, cube_job)

  File "F:\Python\Production\tm1_functions.py", line 283, in extract_data_base_view

    df = tm1.cells.execute_mdx_dataframe(mdx=mdxstr)

  File "F:\Python\lib\site-packages\TM1py\Utils\[Utils.py](https://utils.py/)", line 61, in wrapper

    return func(self, *args, **kwargs)

  File "F:\Python\lib\site-packages\TM1py\Services\[CellService.py](https://cellservice.py/)", line 1697, in execute_mdx_dataframe

    return self.extract_cellset_dataframe(cellset_id, top=top, skip=skip, skip_zeros=skip_zeros,

  File "F:\Python\lib\site-packages\TM1py\Utils\[Utils.py](https://utils.py/)", line 61, in wrapper

    return func(self, *args, **kwargs)

  File "F:\Python\lib\site-packages\TM1py\Services\[CellService.py](https://cellservice.py/)", line 2783, in extract_cellset_dataframe

    raw_csv = self.extract_cellset_csv(

  File "F:\Python\lib\site-packages\TM1py\Services\[CellService.py](https://cellservice.py/)", line 2574, in extract_cellset_csv

    cellset_dict = self.extract_cellset_raw(cellset_id, cell_properties=["Value"], top=top, skip=skip,

  File "F:\Python\lib\site-packages\TM1py\Services\[CellService.py](https://cellservice.py/)", line 50, in wrapper

    return func(self, cellset_id, *args, **kwargs)

  File "F:\Python\lib\site-packages\TM1py\Services\[CellService.py](https://cellservice.py/)", line 2271, in extract_cellset_raw

    cellset_response = self.extract_cellset_raw_response(

  File "F:\Python\lib\site-packages\TM1py\Services\[CellService.py](https://cellservice.py/)", line 2233, in extract_cellset_raw_response

    response = self._rest.GET(url=url, **kwargs)

  File "F:\Python\lib\site-packages\TM1py\Services\[RestService.py](https://restservice.py/)", line 80, in wrapper

    response = self.retrieve_async_response(async_id)

  File "F:\Python\lib\site-packages\TM1py\Services\[RestService.py](https://restservice.py/)", line 590, in retrieve_async_response

    return self._s.get(url, **kwargs)

  File "F:\Python\lib\site-packages\requests\[sessions.py](https://sessions.py/)", line 600, in get

    return self.request("GET", url, **kwargs)

  File "F:\Python\lib\site-packages\requests\[sessions.py](https://sessions.py/)", line 587, in request

    resp = self.send(prep, **send_kwargs)

  File "F:\Python\lib\site-packages\requests\[sessions.py](https://sessions.py/)", line 745, in send

    r.content

  File "F:\Python\lib\site-packages\requests\[models.py](https://models.py/)", line 899, in content

    self._content = b"".join(self.iter_content(CONTENT_CHUNK_SIZE)) or b""

  File "F:\Python\lib\site-packages\requests\[models.py](https://models.py/)", line 818, in generate raise ChunkedEncodingError(e) requests.exceptions.ChunkedEncodingError: ('Connection broken: IncompleteRead(0 bytes read, 2 more expected)', IncompleteRead(0 bytes read, 2 more expected)) 
```

**Comment by rkvinoth** [comment]
One quick workaround is to assign `max_threads` to the `ThreadPool`. What I mean by that is send less number of queries at a time to TM1.
If you set `max_threads` to 15, then TM1 will parse those and return the results. The `Threadpool` will spawn new threads as the old one completes. In this case you will not loose out on wait times.

There is also a `timeout` parameter part of all methods in TM1py which you can leverage I guess. I have never worked with cloud so will differ this to someone else.

**Comment by rclapp** [comment]
You could also look at enabling async mode in tm1py. This isn't uncommon as http connections are not intended to be kept open a long period of time and the can drop as you see. The sync method keeps tm1py from waiting for a response and instead has it periodically check the status .

Sent from my mobile device

On Apr 6, 2023 6:59 PM, Vinoth Kumar Ravi ***@***.***> wrote:

One quick workaround is to assign max_threads to the ThreadPool. What I mean by that is send less number of queries at a time to TM1.
If you set max_threads to 15, then TM1 will parse those and return the results. The Threadpool will spawn new threads as the old one completes. In this case you will not loose out on wait times.

There is also a timeout parameter part of all methods in TM1py which you can leverage I guess. I have never worked with cloud so will differ this to someone else.

**Comment by MariusWirtz** [comment]
@scotthoopes Once we implement https://github.com/cubewise-code/tm1py/issues/887 TM1py will allow you to pass the `max_workers` argument with a large MDX query to the `execute_mdx_dataframe function`. Then TM1py will handle the multi-threaded queries and return the results of the sub-queries as a large combined data frame.

Then you wouldn't need to bother with the multithreading yourself as a user of TM1py. 
Would that help you? Or would you still want to call the functions in parallel yourself?

**Comment by scotthoopes** [comment]
@MariusWirtz reviewing #887 I think that would work great for the use case I have. Right now if I try and do the full view it would take up a lot of resources on the server and PAW won't even load it. Right now I am breaking down each MDX by two elements taking what would be a few millions rows per MDX to 200K to 500K rows. 

In the case of this #887 are you able determine how it breaks down the threads? Or will TM1Py be able to get a large MDX dataset and parse it down to manageable chunks? 

In the end if TM1py could run it in parallel for me that would be great and I would use it. Depending on how it works I probably might need to break down the threads a bit more for our use case. I took our load time from a CSV load that took 2 hours down to about 30 minutes at the moment. If we can cut that down even more that would be great. 

If you have any questions that I can help with please let me know.

**Comment by MariusWirtz** [comment]
For #887, I imagine that we pass an mdxpy object into the `execute_mdx_dataframe` function as an argument instead of a string. That would give TM1py sufficient flexibility to break the query into sub-queries without needing to parse and understand the raw MDX string, which would be tricky.

I think using one dimension to slice would be my preferred option.
So the user passes the `slice_dimension` (e.g. `month`) and `slice_size` (e.g. `3`) as an argument to the function along with `max_workers` (e.g. `4`).
Then TM1py creates subqueries based on the base query. Each sub-query uses a slize of n elements from slice dimension.
They would be executed in parallel and the function returns a combined data frame.

``` python
query = MdxBuilder.from_cube(CUBE)
query.columns_non_empty()
query.add_hierarchy_set_to_column_axis(MdxHierarchySet.all_leaves('Month'))
query.add_hierarchy_set_to_row_axis(MdxHierarchySet.all_leaves(DIM2))

csv = tm1.cells.execute_mdx_csv(query, use_blob=True, max_workers=4, slice_dimension='Month', slice_size=3)
```

**Comment by MariusWirtz** [comment]
resolved with #903
